Revise sample tutorials for ScalarDB Analytics with PostgreSQL, ScalarDB GraphQL, and ScalarDB SQL (JDBC) (#52)
scalar-labs/scalardb-samples: scalardb-graphql-sample/README.md

# Create a Sample Application That Uses ScalarDB GraphQL

This tutorial describes how to create a sample electronic money application that uses ScalarDB GraphQL.

## Overview

> In the sample, you will use Cassandra as the database, and the Cassandra server and the ScalarDB GraphQL server will run in Docker. Then, you will build a Node.js client application that will communicate with the GraphQL server.

Author: We have resolved this in Slack as follows:

- When referring to the ScalarDB GraphQL interface, the proper wording is `ScalarDB GraphQL`.
- When referring to the ScalarDB GraphQL server component, the proper wording is `ScalarDB GraphQL Server`.

```console
$ docker-compose up -d cassandra
```

{% capture notice--info %}
**Note**

Starting the Docker container may take more than one minute depending on your development environment.

To check the logs, run the following command:

```console
$ docker-compose logs -f
```
{% endcapture %}

<div class="notice--info">{{ notice--info | markdownify }}</div>

### Load the schema

The database schema (the method in which the data will be organized) for the sample application has already been defined in [`emoney-schema.json`](emoney-schema.json).

To apply the schema, go to the [ScalarDB Releases](https://github.com/scalar-labs/scalardb/releases) page and download the ScalarDB Schema Loader that matches the version of ScalarDB that you want to use to the `scalardb-samples/scalardb-graphql-sample` folder.

Then, run the following command, replacing `<VERSION>` with the version of the ScalarDB Schema Loader that you downloaded:

```console
$ java -jar scalardb-schema-loader-<VERSION>.jar --config database.properties --schema-file emoney-schema.json --coordinator
```

> 

Author: For consistency with the other sample tutorials (listed below in **Note**), should we have a `### Schema details` section here after the `### Load the schema` section?

If so, what kind of information do we need to include?

> **Note**
> List of sample tutorials that include a `### Schema details` section:
> 
> - [Create a Sample Application by Using ScalarDB](https://scalardb.scalar-labs.com/docs/latest/scalardb-samples/scalardb-sample/README/#schema-details)
> - [Create a Sample Application That Supports Microservice Transactions](https://scalardb.scalar-labs.com/docs/latest/scalardb-samples/microservice-transaction-sample/README/#schema-details)
> - [Create a Sample Application That Supports Multi-Storage Transactions](https://scalardb.scalar-labs.com/docs/latest/scalardb-samples/multi-storage-transaction-sample/README/#schema-details)

Author: @brfrn169 Could you please help me with this part?🙇🏻‍♂️

> ### Start the GraphQL server

Author: For consistency with the other sample tutorials (listed below in **Note**), do we need a `### Load the initial data` section before the `### Start the GraphQL server` section?

If so, what kind of information do we need to include?

> **Note**
> List of sample tutorials that include a `### Schema details` section:
> 
> - [Create a Sample Application by Using ScalarDB](https://scalardb.scalar-labs.com/docs/latest/scalardb-samples/scalardb-sample/README/#load-the-initial-data)
> - [Create a Sample Application That Supports Microservice Transactions](https://scalardb.scalar-labs.com/docs/latest/scalardb-samples/microservice-transaction-sample/README/#load-the-initial-data-by-starting-the-microservices)
> - [Create a Sample Application That Supports Multi-Storage Transactions](https://scalardb.scalar-labs.com/docs/latest/scalardb-samples/multi-storage-transaction-sample/README/#load-the-initial-data)

Author: @brfrn169 Could you please help me with this part?🙇🏻‍♂️

Before starting the GraphQL server, log in to the GitHub Container registry by running the following command and entering your GitHub credentials as instructed:

```console
$ docker login ghcr.io

> ```

Author: I added this part since I needed to run this command in order to run the GraphQL server. Do we need to add any additional information here?

Author: @brfrn169 Could you please help me with this part? Is this new part sufficient?🙇🏻‍♂️

```console
$ node dist/emoney.js charge user1 10
```

To create an account for `user2` that has a balance of `20`, run the following command:

```console
$ node dist/emoney.js charge user2 20
```

### Send electronic money between two accounts

Now that you have created two accounts, you can send funds from one account to the other account.

To have `user1` pay `5` to `user2`, run the following command:

```console
$ node dist/emoney.js pay user1 user2 5
```

### Get an account balance

After sending funds from one account to the other, you can check the balance of each account.

To get the balance of `user1`, run the following command:

```console
$ node dist/emoney.js show user1
```

To get the balance of `user2`, run the following command:

```console
$ node dist/emoney.js show user2
```

### Stop the sample application

To stop the sample application, stop the Docker container by running the following command:

```console
$ docker-compose down
```

## Reference - Application structure

This section describes how this client application is configured in detail.

### Files and directories

The following list is a description of the key files and directories in this application:

> - `graphql`. The GraphQL documents (queries and mutations) for the application functions are located in this directory. The GraphQL Code Generator (`graphql-codegen`) reads the documents to generate the client TypeScript code. For more details, see [GraphQL Code Generator](#graphql-code-generator).

Author: For clarity, can we specify the file name for `@graphql-codegen` as follows?: 

```suggestion
- `graphql`. The GraphQL documents (queries and mutations) for the application functions are located in this directory. The GraphQL Code Generator node module (`node_modules/@graphql-codegen`) reads the documents to generate the client TypeScript code. For more details, see [GraphQL Code Generator](#graphql-code-generator).
```

Reviewer: I think just the package name is good and abstracted well. I'm not sure, but npm might change `node_modules` directory in the future.

Author: Understood, thank you! I'll leave this part as is.

> - `graphql`. The GraphQL documents (queries and mutations) for the application functions are located in this directory. The GraphQL Code Generator (`graphql-codegen`) reads the documents to generate the client TypeScript code. For more details, see [GraphQL Code Generator](#graphql-code-generator).

Reviewer: ```suggestion
- `graphql`. The GraphQL schemas (queries and mutations) for the application functions are located in this directory. The GraphQL Code Generator (`graphql-codegen`) reads the schemas to generate the client TypeScript code. For more details, see [GraphQL Code Generator](#graphql-code-generator).
```
According to https://graphql.org/learn/schema/, "schemas" sounds better to me instead of "documents". But I'm not familiar with GraphQL, so any feedback is welcome

Author: Ah, that's good to know! I kept the wording `document` as is since I found the following reference: [GraphQL Document](https://spec.graphql.org/June2018/#:~:text=A%20GraphQL%20Document%20describes%20a,of%20a%20GraphQL%20type%20system). However, that reference seems a bit outdated (2018).

If `schema` is more commonly used these days, I'm open to changing the wording👍

- `src`. All TypeScript code is located in this directory.

>   - `src/generated/graphql.ts`. This is generated by using `graphql-codegen` and is based on the GraphQL endpoint and the local documents in the `graphql` directory.

Author: Similar to my previous comment, can we specify the file name for `@graphql-codegen` as follows?:

```suggestion
  - `src/generated/graphql.ts`. This is generated by using `node_modules/@graphql-codegen` and is based on the GraphQL endpoint and the local documents in the `graphql` directory.
```

Author: As discussed in a [previous comment](https://github.com/scalar-labs/scalardb-samples/pull/52/files#r1372515620), we'll leave this as is.

  - `src/emoney.ts`. This is the main file of the client application.

> - `codegen.yml`. This is the configuration file for `graphql-codegen`.

Author: Similar to my previous comment, can we specify the file name for `@graphql-codegen` as follows?:

```suggestion
- `codegen.yml`. This is the configuration file for `node_modules/@graphql-codegen`.
```

Author: As discussed in a [previous comment](https://github.com/scalar-labs/scalardb-samples/pull/52/files#r1372515620), we'll leave this as is.

# Create a Sample Application That Uses ScalarDB GraphQL

{% capture notice--info %}
**Note**

> ScalarDB GraphQL is now deprecated. Please see [Getting Started with ScalarDB Cluster](https://scalardb.scalar-labs.com/docs/latest/scalardb-cluster/getting-started-with-scalardb-cluster/) instead.

Reviewer 2: Maybe it would be appropriate to state here: ScalarDB GraphQL Server is now deprecated, and consequently, this sample code is also deprecated. To use ScalarDB GraphQL, you need to use ScalarDB Cluster (enterprise only). For more information about ScalarDB Cluster, See [ScalarDB Cluster document]().

Also, the `scalardb-cluster` repository is private. So maybe we need to link to the doc site?

Author: Thank you! I agree that a bit more information would be helpful here. How about the following?:

```suggestion
ScalarDB GraphQL Server is now deprecated, and consequently, this sample code is also deprecated. To use the ScalarDB GraphQL interface, you need to use ScalarDB Cluster, which is available only in the Enterprise edition. For more information, see [ScalarDB Cluster](https://scalardb.scalar-labs.com/docs/latest/scalardb-cluster/).
```

> [!NOTE]
>
> The link in this case points to the docs site, so I think we should be OK.

Reviewer 2: LGTM!
